### PR TITLE
FLUID-6234: Respect relay propagation directives when sorting application of init model values

### PR DESCRIPTION
**Pull request by amb26**

### CODE REVIEW

**Comment by cindyli** [comment]
This pull request looks great to me. 

@colinbdclark, pass it on to you for a review and a test to ensure it does solve your problem. Thanks.

File: src/framework/core/js/DataBinding.js [hunk 1]
```diff
@@ -329,11 +385,47 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
...
+     * @param transacs {Object} Hash of component id to local ChangeApplier transaction
+     * @param mrec {Object} Hash of component id to enlisted component record
+     * @return Object Hash of component id to list of enlisted component record
+     */
+    fluid.computeInitialOutArcs = function (transacs, mrec) {
+        return fluid.transform(mrec, function (recel, id) {
+            var oneOutArcs = {};
+            var listeners = recel.that.applier.listeners.sortedListeners;
+            fluid.each(listeners, function (listener) {
+                console.log("oIT got listener ", listener);
```
Comment (colinbdclark) [inline src/framework/core/js/DataBinding.js]: Can we remove this <code>console.log</code>? It appears to be leftover from debugging and leaves some extra noise in the console when running an Infusion application.

File: src/framework/core/js/DataBinding.js [hunk 2]
```diff
@@ -329,11 +385,47 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
...
+                if (listener.isRelay && !fluid.isExcludedChangeSource(transacs[id], listener.cond)) {
+                    var targetId = listener.targetId;
+                    if (targetId !== id) {
+                        oneOutArcs[targetId] = true;
+                    }
+                    console.log("Unexcluded arc to ", targetId, " with record ", mrec[targetId]);
```
Comment (colinbdclark) [inline src/framework/core/js/DataBinding.js]: As above—remove logging if possible.

File: src/framework/core/js/DataBinding.js [hunk 3]
```diff
@@ -329,11 +385,47 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
...
+                }
+            });
+            var oneOutArcList = Object.keys(oneOutArcs);
+            var togo = oneOutArcList.map(function (id) {
```
Comment (colinbdclark) [inline src/framework/core/js/DataBinding.js]: So modern! ;)

File: src/framework/core/js/DataBinding.js [hunk 4]
```diff
@@ -345,8 +437,29 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
...
         // at least can't be performed in reverse order ("FLUID-3674 event coordination test" will fail) - need more cases
-        var recs = fluid.values(mrec).sort(fluid.sortCompleteLast);
-        fluid.each(recs, function (recel) {
+
+        // Compute the graph of init transaction relays for FLUID-6234 - one day we will have to do better than this, since there
+        // may be finer structure than per-component - it may be that each piece of model area participates in this relation
+        // differently. But this will require even more ambitious work such as fragmenting all the initial model values along
+        // these boundaries.
+        var outArcs = fluid.computeInitialOutArcs(transacs, mrec);
+        console.log("Got outArcs of ", outArcs);
```
Comment (colinbdclark) [inline src/framework/core/js/DataBinding.js]: As above, remove <code>console.log</code>.

File: src/framework/core/js/DataBinding.js [hunk 5]
```diff
@@ -541,12 +669,33 @@ var fluid_3_0_0 = fluid_3_0_0 || {};
     // required within the "overall" transaction whilst genuine (external) changes continue to arrive.
 
     // TODO: Vast overcomplication and generation of closure garbage. SURELY we should be able to convert this into an externalised, arg-ist form
+    /** Registers a listener operating one leg of a model relay relation, connecting the source and target. Called once or twice from `fluid.connectModelRelay` -
```
Comment (colinbdclark) [inline src/framework/core/js/DataBinding.js]: Thanks for adding this extra documentation!

File: tests/framework-tests/core/js/DataBindingTests.js [hunk 6]
```diff
@@ -56,6 +56,120 @@ https://github.com/fluid-project/infusion/raw/master/Infusion-LICENSE.txt
             }
         });
     };
+
+    fluid.tarzanTests = [{
```
Comment (colinbdclark) [inline tests/framework-tests/core/js/DataBindingTests.js]: Did you mean <code>tarjanTests</code>, or is this a joke?

File: src/framework/core/js/DataBinding.js (hunk 2 above)
Comment (amb26) [reply src/framework/core/js/DataBinding.js]: Sorry to miss these stray logs, now removed

File: tests/framework-tests/core/js/DataBindingTests.js (hunk 6 above)
Comment (amb26) [reply tests/framework-tests/core/js/DataBindingTests.js]: I am sorry to report that this is a form of ioke.

**Pull request merged**